Adding unit tests and modifying PowerShell scripts to pass unit tests.

deannastanley opened the pull request:
Added tests for everything.
Created MSFT_xMySqlUtilities.psm1 to hold all of the functions used by the various Get, Set, and Test functions and added it as a nested module.

Changes to xMySqlProvision.Schema.psm1:
Removed ServiceName as an input and using the default instead.
Added validation to several inputs.
Added Port as an input so you don't have to use the default port.
Added PermissionType so you can specify what permissions you want to give to the user.
Added MySQLVersion as an input parameter (and removed the version constant in the other modules) so that you can manage whatever version of MySQL you want.
Updated the ProductId and Name for installing the MySQL Installer Console.
Updated the inputs to the various module calls to account for added or removed inputs.
Removed references to WordPress.

Changes to the rest of the modules:
Updated the schema file to reflect the changes to the inputs.
Removed the version constants.
Moved helper functions to MSFT_xMySqlUtilities.psm1
Modified the inputs to have more descriptive names and added validation.
Changed the MySQL commands so that they work and added some error handling. Also streamlined some of the functionality.
Removed [CmdletBinding(SupportsShouldProcess=$true)] since it wasn't implemented.

deannastanley committed:
Adding unit tests and modifying PowerShell scripts to pass unit tests.

msftclas commented:
Hi __@deannastanley__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <p>
        It looks like you're a Microsoft contributor (Deanna Stanley (Teksystems)). If you're full-time, we DON'T require a Contribution License Agreement. If you are a vendor, please DO sign the electronic Contribution License Agreement. It will take 2 minutes and there's no faxing! https://cla.microsoft.com.
    </p>

TTYL, MSBOT;

msftclas commented:
__@deannastanley__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, MSBOT;

deannastanley committed:
Type(fix test): Fixed AppVeyor Failures - Changed a tab to spaces, and fixed a couple tests
I had a couple of tests that were failing because I checked for a mock to be called twice and after a code change it was only checked once. I had a couple of tests that were failing because the return value I was sending to the mock was not the right type.  I fixed all of those.

tysonjhayes commented on xMySql.psd1:
```diff
@@ -1,6 +1,6 @@
 @{
 # Version number of this module.
-ModuleVersion = '1.1.0.0'
+ModuleVersion = '1.0.0.1'
```
Revert this please, we'll iterate the version number when we publish to the master.

tysonjhayes commented:
@TravisEz13 @KarolKaczmarek - Hello gents, if you have some time to review this and get it merged we'd appreciate it.

deannastanley committed:
Type(fix): Reverted version number
Put the version number in xMySql.psd1 back to 1.0.0.0.

NarineM commented on DscResources/MSFT_xMySqlDatabase/MSFT_xMySqlDatabase.schema.mof:
```diff
@@ -1,9 +1,10 @@
-    [Key, Description("Name of the database.")] String Name;
-    [Required, Description("Should the database be present or absent."), ValueMap{"Present","Absent"}, Values{"Present","Absent"}] String Ensure;
-    [Required, Description("The root credential that is used to install MySql server."), EmbeddedInstance("MSFT_Credential")] String ConnectionCredential;
+    [Key, Description("Name of the database.")] String DatabaseName;
```
Resource name is xMySqlDatabase , so the property "Name" would mean the database name ( this resource does not use or refer anyhow to any other "name" ) , the property 's  description also says so : "Name of the database", why would renaming Name to be DatabaseName be needed? Please justify. Otherwise, I would suggest to leave the property name as it is.

deannastanley replied:
It seemed clearer to use DatabaseName.  It also makes it DatabaseName throughout the xMySql module.  Since it didn't work the way it was written, I didn't worry about breaking existing configurations.

tysonjhayes replied:
Sure this will break the existing users of this, but personally I think this is better. It consitently names all references to the database regardless of resource being referenced. So all references to the database would end up being DatabaseName, name while generic could refer to any number of things in any of the resources.

NarineM replied:
I m not sure if I follow. Why would it be ok to break existing resources that depend on this resource or existing configurations unless it is really needed. Please justify why renaming the properties are required? I believe the property "Name" for the SQLDatabase resource is descriptive enough and it means the name of the database.

deannastanley replied:
The resource didn't work before.  There are other changes that will break existing resources.  It is better to have clear names for things.  Also if you look at resources such as MSFT_GroupResource they also use a more descriptive name.

NarineM commented on DscResources/MSFT_xMySqlDatabase/MSFT_xMySqlDatabase.psm1:
```diff
@@ -51,120 +8,119 @@ function Get-TargetResource
+    }
+  
+    $arguments = "--execute=SELECT IF(EXISTS (SELECT SCHEMA_NAME FROM INFORMATION_SCHEMA.SCHEMATA WHERE SCHEMA_NAME = '$DatabaseName'), 'Yes','No')", `
+        "--user=root", "--password=$($RootCredential.GetNetworkCredential().Password)", "--port=$(Get-MySqlPort -MySqlVersion $MySqlVersion)", "--silent"
```
I do not seem to find where Get-MySqlPort helper is, same for Get-MySqlExe?

tysonjhayes replied:
Those would be in MSFT_xMySqlUtilities.psm1

NarineM replied:
Thanks. Should not there be an include that imports Utilities helper in the module psm1?

deannastanley replied:
It is listed as a NestedModule in xMySql.psd1.

NarineM replied:
Thanks, just saw that change . It is hard to review a commit that has too many changes, so thanks for pointing to the file.

tysonjhayes replied:
No problem, it's understandable. We know it's a large change we're asking for but we had to add a lot of functionality to get this under tests and incorporate some of the changes we wanted.

NarineM commented on DscResources/MSFT_xMySqlGrant/MSFT_xMySqlGrant.psm1:
```diff
@@ -22,236 +15,166 @@ GrantDoesNotExist=A user with the name {0} does not exist.
         [ValidateSet("ALL PRIVILEGES", "CREATE", "DROP", "DELETE", "INSERT", "SELECT", "UPDATE", "EXECUTE")]
-        [string] $PermissionType = "ALL PRIVILEGES"
+        [ValidateNotNullOrEmpty()]
+        [string] $PermissionType,
```
Why the default for PermissionType is removed?

deannastanley replied:
It seemed like defaulting to ALL PRIVELEGES is not very secure.  I think you should have to specify if you want that level of permissions.

tysonjhayes replied:
In my opinon (Deanna can chime in here) I think asking the user to specify the permission level is the better model instead of assuming we know best and granting all users all privileges. Granting the least amount of permissions from a security context should be the recommendation, asking the operator to be mindful of that makes sense.

That being said, if we can specify multiple permission types, (ie: create and drop, but not delete) we should probably swap this to a string array.

deannastanley replied:
Making it a string array is I believe on my list of things I would like to change in my next round of changes.

NarineM replied:
Yes, my comment was about having a default value for this property (not defaulting it to  ALL PRIVILEGES). Please update this resource to grant the least amount of permissions (instead of making it a required property).

deannastanley replied:
The least amount of permissions depends on what you want the user to be able to do.

NarineM replied:
I agree with you here. I was thinking giving "read" only access by default, but I would take it back. I think it is better to require the resource instance to specify the permission type. Looking at how users are created in my sql, they can get created with no permission whatsoever , then granted to specific permission later on.

NarineM commented on DscResources/MSFT_xMySqlGrant/MSFT_xMySqlGrant.psm1:
```diff
@@ -22,236 +15,166 @@ GrantDoesNotExist=A user with the name {0} does not exist.
+        "--port=$(Get-MySqlPort -MySqlVersion $MySqlVersion)", "--silent"
+    $results = Invoke-MySqlCommand -CommandPath $(Get-MySqlExe -MySqlVersion $MySqlVersion) -Arguments $arguments 2>$ErrorPath
+    
+    Read-ErrorFile -ErrorFilePath $ErrorPath
```
Can you point me where Read-ErrorFile helper is?

tysonjhayes replied:
MSFT_xMySqlUtilities.psm1

NarineM commented on DscResources/MSFT_xMySqlGrant/MSFT_xMySqlGrant.schema.mof:
```diff
@@ -2,12 +2,12 @@
+    [Key, Description("Name of MySQL user.")] String UserName;
+    [Key, Description("MySql database name to grant permissions.")] String DatabaseName;
+    [Required, Description("MySql connection credential used for the root."), EmbeddedInstance("MSFT_Credential")] String RootCredential;
+    [Key, Description("MySql user permission type."), ValueMap{"ALL PRIVILEGES", "CREATE", "DROP", "DELETE", "INSERT", "SELECT", "UPDATE", "EXECUTE"}, Values{"ALL PRIVILEGES", "CREATE", "DROP", "DELETE", "INSERT", "SELECT", "UPDATE", "EXECUTE"}] String PermissionType;
```
Why PermisionType is a key property? 
Why SqlVersion does not have a default value such as the latest MySQL version tested with this resource ?

tysonjhayes replied:
If we specified a default value for the version we would then have to constantly update this resource as the new MySQL version came out, we would also then force people to upgrade to the latest version if they used the default value. Forcing the user to specify the version they would want strikes me as a better idea as upgrades or non-compliance to a version should be driven from the user.

deannastanley replied:
I made PermissionType a key property because you need to call xMySqlGrant multiple times if you want to grant multiple permission types and you can't put those all in the same config if you use just UserName and DatabaseName as key properties.  Adding PermissionType makes a unique key.
It seemed to me like you should know what version you are working with.  Plus then your config wouldn't be broken when the next revision of the module changes the default version.

tysonjhayes replied:
@deannastanley would need to chime in about the PermissionType but my guess is we were using the PowerShellOrg schema creator and that's what it sets to if we don't specify a default property and it contains only one value. As I previously suggested we may want to change this to an array it would make sense to modify this back to write.

deannastanley replied:
Even if you change it to an array, you still probably want to have it as a key property to handle the case where you want to ensure some permissions are set and others are not in the same configuration (like if you are changing permissions for a user).

NarineM replied:
There is no guarantee that this resource will work with upcoming new releases of the installer. I think it should default to the latest version that was tested with this resource and not require the version to be a required property.

deannastanley replied:
Not sure what version you are referring to as your original comment here was about PermissionType.

NarineM commented on DscResources/MSFT_xMySqlServer/MSFT_xMySqlServer.schema.mof:
```diff
@@ -2,19 +2,10 @@
-  EmbeddedInstance("MSFT_Credential"),
-  Description("The root credential that is used to install mySql server.")]
-  String RootPassword;
+    [Key, Description("mySql Version Number")] String MySqlVersion;
```
I agree that adding version support is important , but this change has a solution for it from a wrong angle I believe. Original version of the resource was using ServiceName as a key property as on the same machine you can run multiple MySQL servers of the same version if you install them as services using unique names.

deannastanley replied:
You can still run multiple versions.  It just uses the default service name for them.  Version 5.6.x use MySQL56 and version 5.7.x use MySQL57.

NarineM replied:
It is not about running different versions of the MySQL as a service that use the default service name during the installation, but about installing multiple instances of the same MySQL version and running them under different service names. However, not a blocker.

deannastanley replied:
I hadn't realized you could do that with the MySQL Installer Console.  Supporting multiple instances of the same version would definitely complicate things and require some major code changes.  Can you create two services with the same serviceName?

NarineM commented on DscResources/MSFT_xMySqlServer/MSFT_xMySqlServer.psm1:
```diff
@@ -1,303 +1,112 @@
+            Write-Verbose "Installing MySQL..."
+            $arguments = "community", "install", `
+                "server;$MySqlVersion;$(Get-ArchitectureName):*:servertype=Server;port=$Port;passwd=$($RootPassword.GetNetworkCredential().Password)", "-silent"
+            $null = Invoke-MySqlCommand -CommandPath $(Get-MySqlInstallerConsole) -Arguments $arguments
```
It looks like with this change we are only "copying" or installing the exe of MySQL server (that what the installer does if you do not install mysql as a service) for a given my sql version. With the older version of the resource you could install and run multiple instances of MySQL as service.

deannastanley replied:
This does a full install including setting up the service.

NarineM replied:
What I meant was, it would only install one service using the installer's default service name for a given version. With this change you would not be able to install multiple instances of the same version and run them as different services at the same time as it used to do before.

deannastanley replied:
Again, I was not aware that you could install multiple instances of the same version.  I think it would take a lot of code change to support that.

NarineM replied:
Please see my comment that has the summary of overall changes at the very buttom.

NarineM commented on DscResources/MSFT_xMySqlUser/MSFT_xMySqlUser.schema.mof:
```diff
@@ -2,11 +2,11 @@
+    [Required, Description("Credential for MySQL user."), EmbeddedInstance("MSFT_Credential")] String UserCredential;
+    [Required, Description("MySQL root credential used to create a user."), EmbeddedInstance("MSFT_Credential")] String RootCredential;
+    [Write, Description("Ensure MySQL user is present or absent."), ValueMap{"Present","Absent"}, Values{"Present","Absent"}] String Ensure;
+    [Required, Description("MySQL Version Number")] String MySqlVersion;
```
Again I do not see any benefit of repeating the resource name in the property names.
I also do not think the MySQL version should be  a required property for MySQLUser resource. What are the scenarios (if there even are any) where you would run multiple versions of MySQL server on the same machine and you need to create a user with the same name on both of the servers? At the time you create a user you already have MySQL installed, so the version can be programmatically calculated. I think the parameter mySqlVersion should be optional and default to my sql version found first on the machine during look up if the configuration script does not specify any version.

deannastanley replied:
Renaming all 3 properties makes the code clearer.

There probably aren't any scenarios where you would run multiple versions and want to add the same user to both (and if you do you can add them separately).  That's why I added the version as a required parameter.  If you have multiple versions installed on the same machine, you want to be sure which one you want to be sure which one you're adding the user to.

NarineM replied:
If it is only one version of my sql server installed, then why would you need the version to be a required property for the user resource?

kusuriya replied:
If you have several setups though with different resource setups and different MySQL versions you would need it to verify that you have the correct versions installed with each application group, also on the flipside if you update the resource you would not want your version of MySQL to accidentally change since you are updating your resources to keep up with potential bugs and security issues.  
so as an example my resource specifies a default of MySQL 5.6.13 which is fine because when I made my config files my app requires 5.6.10 - 5.6.13. Later on the resource updates for 5.7 and changes the default to 5.7, I update my resource files and now my SQL server goes to 5.7 and my app happens to break on MySQL 5.7. If I started by making the resource say it explicitly needed 5.6.13 I would not get bit by that. So why add a bug that has no real reason?

NarineM replied:
@kusuriya , yes I agree with you on requiring MySqlVersion property but in the sql server installer resource only. Please see my summary on the overall commit which lists individual resources and proposed updates.

NarineM commented on MSFT_xMySqlUtilities.psm1:
```diff
@@ -0,0 +1,315 @@
+function Invoke-MySqlCommand
```
MySqlUtilities.psm1 looks good.

kusuriya commented:
The version really should remain as it is, a required property. Things using MySQL expect certain versions of MySQL to be installed and you cant make assumptions about that otherwise you would break the app relying on MySQL. Since we have no way of being sure what the matched version of MySQL is we need to make sure the user tells us what the matched version is. Assuming that users will always want to install the latest stable available is a bad assumption, and if there are resources that make the assumption that this resource knows what they want they need to be fixed. It is bad practice all the way around to make version assumptions for users when the versions of the software vary so wildly.

NarineM commented:
I think you are confusing two different things. One thing is specifying the version in the resource instance, the other making it required vs optional. I agree that there should a  version property for my sql server resource, and default to the latest  (well known installer) version tested with this resource. You do not want to end up in the situation where someone is using this resource with the newer version if the installer which does not work with the resource or requires changes in the resource implementation to support newer installer.

kusuriya commented:
No I'm not confusing them, what I'm saying is setting it as a default adds an unwanted behavior that is not needed.

NarineM commented:
Here is the summary on the changes I propose to make for this commit:
xMySqlServer resource:
1. Add a new property MySqlVersion as a key property as long as the installer can be identified by the version
2. Make ServiceName property as optional. If it is not specified , the resource would not pass any servicename to the installer (in that case installer will use the default MySQL<version> service name) and if specified, the resource would pass the service name value to the installer
3. Do not change existing property names unless you justify what renaming them is required. Otherwise it is redundant change.

For all other resources in this module except xMySqlServer:
1. Do not rename the existing property names unless justified why the change is needed.
2. Add a new property MySqlVersion as optional. If the version property is not specified and there are multiple versions of my sql server found in the system, resource should throw an error. If this property is not specified and there is only on version of sql server installed, that version will be used by default
If the property is specified , but there is no my sql server found installed that has the specified version, throw an error, otherwise use the version of my sql server that is specified in the resource instance.
3. Generic error from  all the code paths that anyhow use or depend on MySQL version and given version is not found in the system.
